src/algebra/commute.lean

```diff
@@ -0,0 +1,250 @@
+/-
+Copyright (c) 2019 Neil Strickland. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Neil Strickland
+
+Commuting pairs of elements in not-necessarily-commutative monoids,
+groups and (semi) rings.  Centralizers as submonoids, subgroups or
+subrings.
+
+Usage example: suppose that a, b and c are elements of a ring, and
+that `hb : commute a b` and `hc : commute a c`.  Then
+`hb.pow_left 5` proves `commute (a ^ 5) b` and
+`(hb.pow 2).add (hb.mul hc)` proves `commute a (b ^ 2 + b * c)`.
+Lean does not immediately recognise these terms as equations,
+so for rewriting we need syntax like `rw [(hb.pow_left 5).eq]`
+rather than just `rw [hb.pow_left 5]`.
+-/
+
+import algebra.group_power
+  group_theory.submonoid group_theory.subgroup
+  ring_theory.subring
+
+section semigroup
+
+variables {S : Type*} [semigroup S]
+
+@[reducible] def commute (a b : S) : Prop := a * b = b * a
+
+theorem commute.eq {a b : S} (h : commute a b) : a * b = b * a := h
+
+@[refl] theorem commute.refl (a : S) : commute a a := by { unfold commute }
+
+@[symm] theorem commute.symm {a b : S} : commute a b → commute b a :=
+λ h, h.symm
+
+theorem commute.mul {a b c : S} :
```

Comment:
Shouldn't this be called `mul_right` for consistency?

Reply:
Debatable.  The symmetry is already broken by the use of `commute a x` in definition of centralizers.    Everything would become more verbose if we added `_right` everywhere.

Reply:
@ChrisHughes24 Your suggested changes were already discussed here. I think I agree with Neil that the symmetry is already somewhat broken, and stuff does become more verbose...
I don't have a strong opinion here.

Reply:
@ChrisHughes24 Previous discussion was here.

Reply:
@ChrisHughes24 I have no strong opinion on this matter, and it seems that @NeilStrickland is no longer interested in getting this PR merged. So, I think that you and @jcommelin can decide; I'm ready to rename, if needed.

Reply:
Sorry, only just saw this conversation, as mergify caused notification issues. I still prefer the `left` and `right` names, then it's completely obvious that `mul_right` is the correct lemma, if `mul_left` was the wrong way around. The names `mul` and `mul_right` don't make it obvious that the two lemmas are basically the same.

Reply:
Done.